src/backend/vulkan/src/device.rs

```diff
@@ -365,428 +888,21 @@ impl d::Device<B> for Device {
-        let mut info_viewport_states = Vec::with_capacity(descs.len());
-        let mut info_rasterization_states = Vec::with_capacity(descs.len());
-        let mut info_multisample_states = Vec::with_capacity(descs.len());
-        let mut info_depth_stencil_states = Vec::with_capacity(descs.len());
-        let mut info_color_blend_states = Vec::with_capacity(descs.len());
-        let mut info_dynamic_states = Vec::with_capacity(descs.len());
-        let mut color_attachments = Vec::with_capacity(descs.len());
-        let mut info_specializations = Vec::with_capacity(descs.len() * NUM_STAGES);
-        let mut specialization_entries = Vec::with_capacity(descs.len() * NUM_STAGES);
-        let mut dynamic_states = Vec::with_capacity(descs.len() * MAX_DYNAMIC_STATES);
-        let mut viewports = Vec::with_capacity(descs.len());
-        let mut scissors = Vec::with_capacity(descs.len());
-        let mut sample_masks = Vec::with_capacity(descs.len());
-
-        let mut c_strings = Vec::new(); // hold the C strings temporarily
-        let mut make_stage = |stage, source: &pso::EntryPoint<'a, B>| {
-            let string = CString::new(source.entry).unwrap();
-            let p_name = string.as_ptr();
-            c_strings.push(string);
-
-            let map_entries = source
-                .specialization
-                .constants
-                .iter()
-                .map(|c| vk::SpecializationMapEntry {
-                    constant_id: c.id,
-                    offset: c.range.start as _,
-                    size: (c.range.end - c.range.start) as _,
-                })
-                .collect::<SmallVec<[_; 4]>>();
-
-            specialization_entries.push(map_entries);
-            let map_entries = specialization_entries.last().unwrap();
-
-            info_specializations.push(vk::SpecializationInfo {
-                map_entry_count: map_entries.len() as _,
-                p_map_entries: map_entries.as_ptr(),
-                data_size: source.specialization.data.len() as _,
-                p_data: source.specialization.data.as_ptr() as _,
-            });
-            let info = info_specializations.last().unwrap();
-
-            vk::PipelineShaderStageCreateInfo {
-                s_type: vk::StructureType::PIPELINE_SHADER_STAGE_CREATE_INFO,
-                p_next: ptr::null(),
-                flags: vk::PipelineShaderStageCreateFlags::empty(),
-                stage,
-                module: source.module.raw,
-                p_name,
-                p_specialization_info: info,
-            }
+        let bufs = if cfg!(debug_assertions) {
+            let descs: Vec<_> = descs.into_iter().collect::<Vec<_>>();
+            debug!(
+                "create_graphics_pipelines {:?}",
+                descs.iter().map(Borrow::borrow).collect::<Vec<_>>(),
+            );
+            GraphicsPipelineInfoBuf::collect_from(self, descs.into_iter())
```

Comment:
perhaps we can just print the info in `descs.into_iter().inspect(|desc| debug!("....{:?}", desc))` instead of looping through them